include/particles/particle.h

```diff
@@ -254,56 +305,60 @@ class Particle : public ParticleBase<Tdim> {
+  Eigen::VectorXd tensor_data(const std::string& property);
 
   //! Apply particle velocity constraints
   //! \param[in] dir Direction of particle velocity constraint
   //! \param[in] velocity Applied particle velocity constraint
-  void apply_particle_velocity_constraints(unsigned dir,
-                                           double velocity) override;
+  void apply_particle_velocity_constraints(unsigned dir, double velocity);
 
   //! Assign material id of this particle to nodes
-  void append_material_id_to_nodes() const override;
+  void append_material_id_to_nodes() const;
 
   //! Return the number of neighbour particles
-  unsigned nneighbours() const override { return neighbours_.size(); };
+  unsigned nneighbours() const { return neighbours_.size(); };
 
   //! Assign neighbour particles
   //! \param[in] neighbours set of id of the neighbouring particles
   //! \retval insertion_status Return the successful addition of a node
-  void assign_neighbours(const std::vector<mpm::Index>& neighbours) override;
+  void assign_neighbours(const std::vector<mpm::Index>& neighbours);
 
   //! Return neighbour ids
-  std::vector<mpm::Index> neighbours() const override { return neighbours_; };
+  std::vector<mpm::Index> neighbours() const { return neighbours_; };
 
  private:
   //! Compute strain rate
   inline Eigen::Matrix<double, 6, 1> compute_strain_rate(
       const Eigen::MatrixXd& dn_dx, unsigned phase) noexcept;
 
  private:
-  //! particle id
-  using ParticleBase<Tdim>::id_;
+  //! Particle id
+  Index id_{std::numeric_limits<Index>::max()};
   //! coordinates
-  using ParticleBase<Tdim>::coordinates_;
-  //! Reference coordinates (in a cell)
-  using ParticleBase<Tdim>::xi_;
-  //! Cell
-  using ParticleBase<Tdim>::cell_;
+  VectorDim coordinates_;
   //! Cell id
-  using ParticleBase<Tdim>::cell_id_;
-  //! Nodes
-  using ParticleBase<Tdim>::nodes_;
+  Index cell_id_{std::numeric_limits<Index>::max()};
   //! Status
-  using ParticleBase<Tdim>::status_;
+  bool status_{true};
+  //! Reference coordinates (in a cell)
+  Eigen::Matrix<double, Tdim, 1> xi_;
+  //! Cell
+  std::shared_ptr<Cell<Tdim>> cell_;
+  //! Vector of nodal pointers
+  std::vector<std::shared_ptr<NodeBase<Tdim>>> nodes_;
   //! Material
-  using ParticleBase<Tdim>::material_;
-  //! Material id
-  using ParticleBase<Tdim>::material_id_;
-  //! State variables
-  using ParticleBase<Tdim>::state_variables_;
-  //! Neighbour particles
-  using ParticleBase<Tdim>::neighbours_;
+  std::shared_ptr<Material<Tdim>> material_;
+  //! Unsigned material id
+  unsigned material_id_{std::numeric_limits<unsigned>::max()};
+  //! Material state history variables
+  mpm::dense_map state_variables_;
+  //! Vector of particle neighbour ids
+  std::vector<mpm::Index> neighbours_;
+  //! Scalar properties
+  tsl::ordered_map<mpm::properties::Scalar, double> scalar_properties_;
+  //! Vector properties
+  tsl::ordered_map<mpm::properties::Scalar, Eigen::Matrix<double, 1, Tdim>>
```

Comment:
Why `mpm::properties::Scalar`? What enum and data type are you planning to define vectorial variables like velocity?

Reply:
I think @kks32 has not made the enum for Vector. The most important thing is that the map has `Eigen::Matrix<double, 1, Tdim>` which is a vector. On that note (I will write it in general comment), do we need tensor too? How do we store stresses?

Reply:
Not at the moment, we'll use stresses and strains  direclty for no